File: backend/import/presto/domain/jobs.py
from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any, Protocol, Literal

from .capabilities import DawTarget
from .errors import PrestoErrorPayload


JobState = Literal["queued", "running", "succeeded", "failed", "cancelled"]


@dataclass(frozen=True)
class JobProgress:
    phase: str
    current: int
    total: int
    percent: float
    message: str | None = None


@dataclass
class JobRecord:
    job_id: str
    capability: str
    target_daw: DawTarget
    state: JobState
    progress: JobProgress = field(default_factory=lambda: JobProgress(phase="", current=0, total=0, percent=0.0))
    metadata: dict[str, Any] | None = None
    result: object | None = None
    error: PrestoErrorPayload | None = None
    created_at: str = ""
    started_at: str | None = None
    finished_at: str | None = None


@dataclass(frozen=True)
class JobAcceptedResponse:
    job_id: str
    capability: str
    state: Literal["queued", "running"]


@dataclass(frozen=True)
class JobsGetResponse:
    job: JobRecord


@dataclass(frozen=True)
class JobsCreateRequest:
    capability: str
    target_daw: DawTarget
    state: JobState | None = None
    progress: dict[str, Any] | None = None
    metadata: dict[str, Any] | None = None
    result: object | None = None
    error: PrestoErrorPayload | None = None
    started_at: str | None = None
    finished_at: str | None = None


@dataclass(frozen=True)
class JobsCreateResponse:
    job: JobRecord


@dataclass(frozen=True)
class JobsListRequest:
    states: tuple[JobState, ...] | None = None
    capabilities: tuple[str, ...] | None = None
    limit: int | None = None


@dataclass(frozen=True)
class JobsListResponse:
    jobs: list[JobRecord]
    total_count: int


@dataclass(frozen=True)
class JobsCancelResponse:
    cancelled: bool
    job_id: str


@dataclass(frozen=True)
class JobsDeleteResponse:
    deleted: bool
    job_id: str


@dataclass(frozen=True)
class JobsUpdateRequest:
    job_id: str
    state: JobState | None = None
    progress: dict[str, Any] | None = None
    metadata: dict[str, Any] | None = None
    result: object | None = None
    error: PrestoErrorPayload | None = None
    started_at: str | None = None
    finished_at: str | None = None


@dataclass(frozen=True)
class JobsUpdateResponse:
    job: JobRecord


class JobManagerProtocol(Protocol):
    def upsert(self, job: JobRecord) -> None:
        ...

    def create(self, request: JobsCreateRequest) -> JobsCreateResponse:
        ...

    def update(self, request: JobsUpdateRequest) -> JobsUpdateResponse:
        ...

    def get(self, job_id: str) -> JobRecord:
        ...

    def list(self, filter: JobsListRequest | None = None) -> JobsListResponse:
        ...

    def cancel(self, job_id: str) -> JobsCancelResponse:
        ...

    def delete(self, job_id: str) -> JobsDeleteResponse:
        ...

File: packages/sdk-core/src/clients/jobs.ts
import type {
  JobRecord,
  JobsCancelRequest,
  JobsCancelResponse,
  JobsClient,
  JobsCreateRequest,
  JobsCreateResponse,
  JobsDeleteRequest,
  JobsDeleteResponse,
  JobsGetRequest,
  JobsGetResponse,
  JobsListRequest,
  JobsListResponse,
  JobsUpdateRequest,
  JobsUpdateResponse,
  PublicCapabilityId,
} from '@presto/contracts'
import type { PrestoClientAssemblyContext } from '../createPrestoClient'

const invokeCapability = async <TRequest, TResponse>(
  context: PrestoClientAssemblyContext,
  capability: PublicCapabilityId,
  payload: TRequest,
): Promise<TResponse> => {
  const response = await context.transport.invoke<TRequest, TResponse>({
    requestId: context.nextRequestId(),
    capability,
    payload,
    meta: {
      clientName: context.clientName,
      clientVersion: context.clientVersion,
    },
  })

  if (response.success === false) {
    throw response.error
  }

  return response.data
}

export const createJobsClient = (context: PrestoClientAssemblyContext): JobsClient => ({
  create: (request: JobsCreateRequest) =>
    invokeCapability<JobsCreateRequest, JobsCreateResponse>(context, 'jobs.create', request),
  update: (request: JobsUpdateRequest) =>
    invokeCapability<JobsUpdateRequest, JobsUpdateResponse>(context, 'jobs.update', request),
  get: async (jobId: string): Promise<JobRecord> => {
    const response = await invokeCapability<JobsGetRequest, JobsGetResponse>(
      context,
      'jobs.get',
      { jobId },
    )

    return response.job
  },
  list: (filter?: JobsListRequest) =>
    invokeCapability<JobsListRequest, JobsListResponse>(
      context,
      'jobs.list',
      filter ?? {},
    ),
  cancel: (jobId: string) =>
    invokeCapability<JobsCancelRequest, JobsCancelResponse>(context, 'jobs.cancel', { jobId }),
  delete: (jobId: string) =>
    invokeCapability<JobsDeleteRequest, JobsDeleteResponse>(context, 'jobs.delete', { jobId }),
})
